HSEARCH-4923 + HSEARCH-4985 Tidy up documentation of listener-triggered indexing vs indexing plans (#3769)
hibernate/hibernate-search: documentation/src/main/asciidoc/public/reference/_indexing-explicit.adoc

[[indexing-explicit]]
= [[mapper-orm-indexing-manual]] [[manual-index-changes]] Explicit indexing

[[indexing-explicit-basics]]
== [[mapper-orm-indexing-manual-basics]] [[search-batchindex]] Basics

While <<listener-triggered-indexing,listener-triggered indexing>> and
the <<indexing-massindexer,`MassIndexer`>>
or <<mapper-orm-indexing-jakarta-batch,the mass indexing job>>
should take care of most needs,
it is sometimes necessary to control indexing manually.

The need arises in particular when <<listener-triggered-indexing,listener-triggered indexing>> is <<mapping-annotations,disabled>>
or simply not supported (e.g. <<mapper-pojo-standalone-indexing-listener-triggered,with the Standalone POJO Mapper>>),

> or when entity changes happen listener-triggered indexing did not or cannot detect --
> <<limitations-changes-in-session,such as JPQL/SQL `insert`, `update` or `delete` queries>>.

Reviewer: ```suggestion
or when listener-triggered indexing cannot detect entity changes happening --
<<limitations-changes-in-session,e.g. JPQL/SQL `insert`, `update` or `delete` queries>>.
```
Maybe something like that? saying `listener-triggered indexing did not detect` feels like it should've but it didn't and  that feels more like  a 🪳  😃

[[listener-triggered-indexing-synchronization]]
== Configuration

As explicit indexing uses <<indexing-plan,indexing plans>> under the hood,
several configuration options affecting indexing plans will affect explicit indexing as well:

* The <<indexing-plan-synchronization,indexing plan synchronization strategy>>.
* The <<indexing-plan-filter,indexing plan filter>>.

[[indexing-explicit-plan]]
== [[mapper-orm-indexing-manual-indexingplan-writes]] [[_deleting_instances_from_the_index]] [[_adding_instances_to_the_index]] Using a `SearchIndexingPlan` manually

Explicit access to the <<indexing-plan,indexing plan>> is done in the context of a <<entrypoints-search-session,`SearchSession`>>
using the `SearchIndexingPlan` interface.
This interface represents the (mutable) set of changes
that are planned in the context of a session,
and will be applied to indexes upon transaction commit (for the <<mapper-orm,Hibernate ORM integration>>)
or upon closing the `SearchSession` (for the <<mapper-pojo-standalone,Standalone POJO Mapper>>).

Here is how explicit indexing based on an <<indexing-plan,indexing plan>> works at a high level:

1. When the application wants an index change,

> its calls one of the `add`/`addOrUpdate`/`delete` methods on the indexing plan of the current <<entrypoints-search-session,`SearchSession`>>.

Reviewer: ```suggestion
it calls one of the `add`/`addOrUpdate`/`delete` methods on the indexing plan of the current <<entrypoints-search-session,`SearchSession`>>.
```

Prefer `addOrUpdate` unless you are really sure of yourself and need a (slight) performance boost.
`addOrUpdate(Object entity)`::
Add or update a document in the index if the entity type is mapped to an index (`@Indexed`),
and re-index documents that embed this entity (through `@IndexedEmbedded` for example).
`delete(Object entity)`::
Delete a document from the index if the entity type is mapped to an index (`@Indexed`),
and re-index documents that embed this entity (through `@IndexedEmbedded` for example).
`purge(Class<?> entityType, Object id)`::
Delete the entity from the index,
but do not try to re-index documents that embed this entity.
+
Compared to `delete`, this is mainly useful if the entity has already been deleted from the database
and is not available, even in a detached state, in the session.
In that case, reindexing associated entities will be the user's responsibility,
since Hibernate Search cannot know which entities are associated to an entity that no longer exists.
`purge(String entityName, Object id)`::
Same as `purge(Class<?> entityType, Object id)`,
but the entity type is referenced by its name (see `@javax.persistence.Entity#name`).

> `process()` and `execute()`::

Reviewer: Maybe it's better to split these into two items? since execute would write to an index in `no-coordination` case and these updates won't be rolled back in case of a transaction rollback...
